internal/rbd/controllerserver.go

```diff
@@ -357,6 +358,56 @@ func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSna
 	return nil
 }
 
+func (cs *ControllerServer) ListSnapshots(
+	ctx context.Context,
+	req *csi.ListSnapshotsRequest,
+) (*csi.ListSnapshotsResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(
+		csi.ControllerServiceCapability_RPC_LIST_SNAPSHOTS); err != nil {
+		log.ErrorLog(ctx, "invalid list snapshot req: %v", protosanitizer.StripSecrets(req))
+
+		return nil, err
+	}
+
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	snap, err := mgr.GetSnapshotByID(ctx, req.GetSnapshotId())
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "failed to find snapshot with ID %q: %s", req.GetSnapshotId(), err.Error())
+	}
+	defer snap.Destroy(ctx)
+
+	// We need type casting here as SourceVolumeID is unset
+	// We fetch the source volume ID from the request and return
+	// that in the response
+	rbdSnap, ok := snap.(*rbdSnapshot)
```

Comment:
The comment isn't explaining a good reason to typecast. `mgr.GetSnapshotByID()` should return a complete Snapshot object. The `SourceVolumeID` in the request is optional, so you can not use that, see the comment I left for CephFS.

Reply:
From the [CSI Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot):

```
// Information about a specific snapshot.
message Snapshot {
  ...

  // Identity information for the source volume. Note that creating a
  // snapshot from a snapshot is not supported here so the source has to
  // be a volume. This field is REQUIRED.
  string source_volume_id = 3;
```

`.SourceVolumeId` is a REQUIRED attribute.

Reply:
Still sounds very dangerous to me. Have you guys checked that after getting and setting the updated fields in the VolumeSnapshot CR, the SourceVolumeId does not get wiped?

When an updated external-snapshotter is used, this will need to be verified again too.

Reply:
Done